draft-ietf-quic-transport.md

```diff
@@ -2427,6 +2407,32 @@ ACK Block (repeated):
   packets preceding the largest packet number, as determined by the
   preceding Gap.
 
+### Sending ACK Frames
+
+Implementations MUST NOT generate packets that only contain ACK frames in
+response to packets which only contain ACK frames. However, they SHOULD
+acknowledge packets containing only ACK frames when sending ACK frames in
+response to other packets.  Implementations MUST NOT send more than one ACK
+frame per received packet that contains frames other than ACK frames.  Packets
+containing non ACK frames MUST be acknowledged immediately or when a delayed
+ack timer expires.
+
+To limit ACK blocks to those that have not yet been received by the sender, the
+receiver SHOULD track which ACK frames have been acknowledged by its peer.  Once
+an ACK frame has been acknowledged, the packets it acknowledges SHOULD NOT be
+acknowledged again.
+
+A receiver that is only sending ACK frames will not receive acknowledgments for
+its packets.  Sending an occasional MAX_DATA or MAX_STREAM_DATA frame as data is
+received will ensure that acknowledgements are generated by a peer.  Otherwise,
+an endpoint MAY send a PING frame once per RTT to solicit an acknowledgment.
+
+To limit receiver state or the size of ACK frames, a receiver MAY limit the
+number of ACK blocks it sends.  A receiver can do this even without receiving
+acknowledgment of its ACK frames, with the knowledge this could cause the sender
+to unnecessarily retransmit some data.  When this is necessary, the receiver
+SHOULD acknowledge newly received packets and stop acknowledging packets
```

Comment:
I think that s/SHOULD/can/  I don't think that we want to mandate anything here.

That is, unless your intent was to say that acknowledging the most recently received packets is preferable, but then I would say "if acknowledgments are removed rather than retransmitted, a receiver SHOULD discard acknowledgments with the lowest packet number first".  Or something along those lines.

Reply:
lowest packet number isn't necessarily the one that was the oldest. What you want really is to discard acks for the earliest received packets. I'm not concerned about the SHOULD here, I think that's fine... but I'm also fine with replacing that with a "can".

Reply:
It's oldest at the sender, so I am assuming that it will be marked as lost first (or might have already).

Separately, should we also note that dropping ACKs causes packets to be marked as lost causes congestion control to receive a warped signal?

Reply:
I think SHOULD is still good here, since this is what we recommend implementations do, and as Jana said, in theory it's not as simple as dropping smallest packet number.

I'll add a note that dropping ack blocks early increases the chance the peer never receives an acknowledgement and may be forced to spuriously retransmit the data it contains.